step_walker variance

mjmadsen opened the pull request:
Switched walk in config to walk_max and walk_min to add variance to stepper.
Also added variance when close to stop/pokemon.
Removed redundant parameters.
Updated configs to reflect change.

Updated for conflicts (#3857).

DiegoRBaquero commented:
@mjmadsen Maybe add a comment that the config is in meters/second? People having speed of 20+ is just a guaranteed ban.

mjmadsen commented:
From what I can tell, the previous code uses walk as km per hour. distance_unit is used purely for output (from what I can tell).

DiegoRBaquero commented:
I just checked and this is with 10 as the walk setting:

`````
2016-08-14 13:14:44,265 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Love Poem - 0.16km
2016-08-14 13:14:46,785 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Love Poem - 0.15km
2016-08-14 13:14:49,615 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Love Poem - 0.14km
2016-08-14 13:14:51,884 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Love Poem - 0.13km
2016-08-14 13:14:55,459 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Love Poem - 0.12km
2016-08-14 13:14:57,573 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Love Poem - 0.11km
2016-08-14 13:15:00,094 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Love Poem - 0.10km
2016-08-14 13:15:02,733 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Love Poem - 0.09km
2016-08-14 13:15:04,415 [IncubateEggs] [INFO] [next_egg_incubates] Next egg incubates in 0.80 km
2016-08-14 13:15:05,297 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Love Poem - 0.08km
`````

and:
`````
2016-08-14 13:17:07,501 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Instituto De Genética UN - 0.15km
2016-08-14 13:17:10,008 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Instituto De Genética UN - 0.14km
2016-08-14 13:17:13,232 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Instituto De Genética UN - 0.13km
2016-08-14 13:17:15,945 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Instituto De Genética UN - 0.12km
2016-08-14 13:17:18,298 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Instituto De Genética UN - 0.11km
2016-08-14 13:17:21,064 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Instituto De Genética UN - 0.10km
2016-08-14 13:17:23,553 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Instituto De Genética UN - 0.09km
2016-08-14 13:17:26,780 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Instituto De Genética UN - 0.08km
2016-08-14 13:17:29,272 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Instituto De Genética UN - 0.07km
2016-08-14 13:17:31,594 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Instituto De Genética UN - 0.05km
2016-08-14 13:17:34,121 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Instituto De Genética UN - 0.04km
````

and
`````
2016-08-14 13:18:18,457 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Mural Abstracto - 0.13km
2016-08-14 13:18:20,780 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Mural Abstracto - 0.12km
2016-08-14 13:18:23,661 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Mural Abstracto - 0.11km
2016-08-14 13:18:26,380 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Mural Abstracto - 0.10km
2016-08-14 13:18:28,486 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Mural Abstracto - 0.08km
2016-08-14 13:18:31,086 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Mural Abstracto - 0.07km
2016-08-14 13:18:33,652 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Mural Abstracto - 0.06km
2016-08-14 13:18:36,887 [MoveToFort] [INFO] [moving_to_fort] Moving towards pokestop Mural Abstracto - 0.05km
`````

Which gives us **13.68 km/h (3.8 m/s)**, **14.65 km/h (4.07 m/s)** and **16 km/ (4.44m/s)** respectively.

We should try and match km/h by diving the given speed in ~1.45 in that case, thoughts?

mjmadsen commented:
@DiegoRBaquero That's interesting. Certainly could explain why my walk = 14 accounts got softbans. I'm not sure why it would be such a difference. The long_lat random is supposed to add a variance of -/+ 9.1 ft, so I don't think that would account for an extra 3-6km/hr.

DiegoRBaquero commented:
@mjmadsen No, that was using 10, no variance at all. The random delta **in average** shouldn't add extra km/h at all.

mjmadsen commented:
Wonder if we have a km to long/lat error somewhere. 

`======================================================================
FAIL: test_normalized_distance (tests.step_walker_test.TestStepWalker)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/PokemonGoF/PokemonGo-Bot/tests/step_walker_test.py", line 33, in test_normalized_distance
    self.assertGreater(sw.dLat, 0)
AssertionError: 0 not greater than 0
======================================================================
FAIL: test_normalized_distance_times_2 (tests.step_walker_test.TestStepWalker)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/PokemonGoF/PokemonGo-Bot/tests/step_walker_test.py", line 44, in test_normalized_distance_times_2
    self.assertTrue(sw.dLat > 0)
AssertionError: False is not true`

Error from last test. I'm going to sleep. I wish these test files weren't part of the main branches...

mjmadsen commented:
@douglascamata The test file hack is less than ideal. But the core code is ultimately more readable because of the changes to parameters. Let me know what you think.

douglascamata commented:
@mjmadsen can you please remove all reference to walk speed being in kilometers per hour? It's an empirical number and thus has no unit.

TheSavior commented:
We want to avoid breaking everyones' config. In this situation, perhaps it makes the most sense to log a warning on startup if they are specifying walk, stating that it is deprecated. If walk is specified, then lets make `walk_max` and `walk_min` set to that.

DiegoRBaquero commented:
@douglascamata Shouldn't we match it to a unit? It makes it easy to set min/max values.

douglascamata commented:
@DiegoRBaquero we can't match it to a unit, because we need to stop walking every now and them to process stuff around us, so it's impossible to have a unit.

DiegoRBaquero commented:
@douglascamata When the ONLY action that runs is walking, as in my comment: https://github.com/PokemonGoF/PokemonGo-Bot/pull/3943#issuecomment-239688410, we can get really near to km/h dividing by 1.45

I can get a more exact average if we will implement the simple division.

TheSavior commented:
The unit is distance, but not velocity. The unit is how far we will move on each step, but we don't make any guarantees as to how many steps per time we make.

keugnu commented:
Maybe this is done better with setting a boolean in the config and varying the existing speed attribute periodically by a random percentage if the bool was set to true. Doesn't break existing configs.

Quantra commented:
With regards to speed would it be possible to measure the time between steps and alter the step distance based on that?  Then if we've spent ages recycling items or w/e it won't matter we'll just make a bigger step next time?

DiegoRBaquero commented:
@TheSavior Then we should add in the doc that the distance is in meters and that a value of 4.16 (default) turns, approximately into 6 km/h.

mjmadsen commented:
@Quantra I don't think the app updates position when in menus, stops or catching pokemon. So we may not want to push our distance change if we've been caught up in those actions for a while.

@TheSavior I get that we can't guarantee distance/hr, but it would be nice if walking speed were directly related to some form of measure or as @DiegoRBaquero mentioned, we indicate an approx. relationship between walk values and km/hr. However, these (imo) are not in the scope of this PR.

I'm at work right now, so I can't test much. However, I'll see about adding a comment on startup about walk in the config. I don't think anyone's config would break with the present code - we still pull default values when the config is loaded.

Quantra commented on pokemongo_bot/step_walker.py:
```diff
@@ -19,18 +20,21 @@ def __init__(self, bot, speed, dest_lat, dest_lng):
 
-        self.speed = speed
+        if self.bot.config.walk == 0:
+            self.speed = self.bot.config.walk_min - random() * (self.bot.config.walk_max - self.bot.config.walk_min)
```
I think this should be
`self.speed = self.bot.config.walk_max - random() * (self.bot.config.walk_max - self.bot.config.walk_min)`
?

mjmadsen replied:
Yep! Thank you.

mjmadsen commented:
I removed the switch to use walk instead of walk_min/walk_max and removed walk from the config. It will generate a warning on startup and will still have the default values for walk_max/walk_min. Issues arose when I tried to implement the conditions on step_walker from the test step_walker file and I don't want to mess around all day with a file that hardly has any use (and could be combined into a single function).

TheSavior commented on pokecli.py:
```diff
@@ -504,7 +514,8 @@ def task_configuration_error(flag_name):
 
     old_flags = ['mode', 'catch_pokemon', 'spin_forts', 'forts_spin', 'hatch_eggs', 'release_pokemon', 'softban_fix',
-                'longer_eggs_first', 'evolve_speed', 'use_lucky_egg', 'item_filter', 'evolve_all', 'evolve_cp_min', 'max_steps']
+                'longer_eggs_first', 'evolve_speed', 'use_lucky_egg', 'item_filter', 'evolve_all', 'evolve_cp_min',
```
We don't want to add it here because this will throw saying that it doesn't exist anymore. You should just check for walk being set, print a message that walk_min and walk_max are used now instead, and then the bot should be able to keep running

mjmadsen replied:
Switch it to a warning below.

TheSavior commented:
:+1:

[![Approved with PullApprove](https://img.shields.io/badge/pullapprove-approved-brightgreen.svg)](https://pullapprove.com/PokemonGoF/PokemonGo-Bot/pull-request/3943/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=PokemonGoF/PokemonGo-Bot)

dezayner commented:
I got this error
[01:09] Walking from (-7.982077060831172, 112.63066117494552, 0) to (-7.982088600000001, 112.6310973, 0) (48.0422623786 m)
[01:09]
[01:09] Ran for 0:01:40
[01:09] Total XP Earned: 110  Average: 3960.12/h
[01:09] Travelled 0.00km
[01:09] Visited 0 stops
[01:09] Encountered 1 pokemon, 1 caught, 1 released, 0 evolved, 0 never seen before
[01:09] Threw 1 pokeball
[01:09] Earned 100 Stardust
[01:09]
[01:09] Highest CP Pokemon: Scyther [CP: 548] [IV: 3/14/12] Potential: 0.64
[01:09] Most Perfect Pokemon: Scyther [CP: 548] [IV: 3/14/12] Potential: 0.64
Traceback (most recent call last):
  File "pokecli.py", line 610, in <module>
    main()
  File "pokecli.py", line 105, in main
    bot.tick()
  File "D:\gamezone\pokebots\PokemonGo-Bot-Dev\pokemongo_bot\__init__.py", line 490, in tick
    if worker.work() == WorkerResult.RUNNING:
  File "D:\gamezone\pokebots\PokemonGo-Bot-Dev\pokemongo_bot\cell_workers\follow_spiral.py", line 117, in work
    if dist <= 1 or (self.bot.config.walk > 0 and step_walker == None):
AttributeError: 'Namespace' object has no attribute 'walk'
